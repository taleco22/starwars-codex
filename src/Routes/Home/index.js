import React from "react";
import {Link} from "react-router-dom";

export default function Home() {
  return (
    <div>
      <h1>Star Wars</h1>
      <Link to="/characters">click me</Link>
    </div>
  );
}
