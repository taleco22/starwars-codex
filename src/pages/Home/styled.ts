import styled from "styled-components";
import { globalButton, globalItem, globalList, globalWindow } from "../../global-styled/index";

export const Window = styled(globalWindow)``;

export const List = styled(globalList)``;
export const Item = styled(globalItem)``;
export const Button = styled(globalButton)``;

export const HomeWrapper = styled.div`
  width: 100%;
  background: blue;
  margin: 0 auto 8rem;
  @media (min-width: 768px) {
    width: 40%;
  }
`;

export const StarLogo = styled.img`
  width: 100%;
  @media (min-width: 768px) {
    width: 50%;
  }
`;
