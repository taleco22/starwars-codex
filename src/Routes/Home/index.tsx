import React, { useEffect, useState } from "react";
import { useHistory } from "react-router-dom";
import logo from "../../logo.svg";
import * as S from "../../styles/styled";
import { initNavigation } from "../../utils/navigation";

const start = () => {
  const audio: HTMLMediaElement | null = document.querySelector("#start");
  audio?.load();
  audio?.play();

  const video: HTMLMediaElement | null = document.querySelector("#video");
  video?.classList.add("play");
  video?.load();
  video?.play();
  video?.addEventListener("ended", () => video.classList.remove("play"));

  return { audio, video };
};

const Home: React.FC = () => {
  const [shake, setShake] = useState(false);

  useEffect(() => {
    initNavigation();
  }, []);

  const history = useHistory();

  const goto = (url: string) => {
    start();
    setShake(true);
    setTimeout(() => {
      setShake(false);
      history.push(url);
    }, 1000);
  };

  return (
    <S.Window shake={shake}>
      <S.StarLogo src={logo} alt="Star Wars" />
      <S.HomeWrapper>
        <S.Button onClick={() => goto("/categories")}>Start</S.Button>
      </S.HomeWrapper>
    </S.Window>
  );
};

export default Home;
