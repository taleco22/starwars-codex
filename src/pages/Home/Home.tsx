import React, { useEffect, useState } from "react";
import { RouteComponentProps } from "react-router-dom";
import logo from "../../logo.svg";
import { waitButtons } from "../../utils/navigation";
import { Button, HomeWrapper, StarLogo, Window } from "./styled";

export const start = () => {
  const audio: HTMLMediaElement | null = document.querySelector("#lightsaberSound");
  audio?.classList.add("play");
  audio?.load();
  audio?.play();
  audio?.addEventListener("ended", () => audio.classList.remove("play"));

  const video: HTMLMediaElement | null = document.querySelector("#lightsaberVideo");
  video?.classList.add("play");
  video?.load();
  video?.play();
  video?.addEventListener("ended", () => video.classList.remove("play"));

  return { audio, video };
};

const Home: React.FC<RouteComponentProps> = ({ history }) => {
  const [shake, setShake] = useState(false);

  useEffect(() => {
    waitButtons();
  }, []);

  const goto = (url: string) => {
    start();
    setShake(true);
    setTimeout(() => {
      setShake(false);
      history.push(url);
    }, 1000);
  };

  return (
    <Window shake={shake}>
      <StarLogo src={logo} alt="Star Wars" />
      <HomeWrapper>
        <Button onClick={() => goto("/categories")}>Start</Button>
      </HomeWrapper>
    </Window>
  );
};

export default Home;
