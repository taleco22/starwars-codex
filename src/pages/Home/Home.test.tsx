// @ts-nocheck
import "@testing-library/jest-dom"
import { fireEvent, render, screen } from "@testing-library/react"
import * as React from "react"
import Home, { playLightsaberFX } from "./Home"

jest.mock("~/utils")
jest.mock("~/assets")

const stateSetter: React.Dispatch<unknown> = jest.fn()

const use_state = jest
  .spyOn(React, "useState")
  .mockImplementation((stateValue) => [stateValue, stateSetter])

const use_effect = jest.spyOn(React, "useEffect")

const history = {
  push: jest.fn((url) => url),
}
const use_history = jest.spyOn(history, "push")

describe("Home.tsx", () => {
  let result: { container: HTMLElement }

  beforeEach(async () => {
    result = render(<Home history={history} />)
  })

  it("render component correctly", () => {
    expect(result.container).toMatchSnapshot()
  })

  it("initialize component", () => {
    expect(use_effect).toBeCalled()
    expect(use_state).toBeCalled()
  })

  it("trigger onClick handler", () => {
    const button = screen.getByText("Start")
    fireEvent.click(button)

    expect(stateSetter).toBeCalledTimes(1)
    expect(setTimeout).toBeCalled()
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 1000)
  })

  jest.useFakeTimers()
  it("waits 1 second to update state and navigate page", () => {
    jest.advanceTimersByTime(1000)

    expect(stateSetter).toBeCalledTimes(2)
    expect(use_history).toBeCalledTimes(1)
  })
})

describe("playLightsaberFX()", () => {
  document.body.innerHTML = `<div><audio id="lightsaberSound"></audio>'+'<video id="lightsaberVideo"><source type='video/mp4; codecs="avc1.42E01E, mp4a.40.2"'/></video></div>`

  it("play sound and video", () => {
    const { audio, video } = playLightsaberFX()
    expect(audio).toHaveClass("play")
    expect(video).toHaveClass("play")
  })

  it("remove class on ended event", () => {
    const { audio, video } = playLightsaberFX()

    audio && fireEvent.ended(audio)
    video && fireEvent.ended(video)

    expect(audio).not.toHaveClass("play")
    expect(video).not.toHaveClass("play")
  })

  it("both audio and video are null", () => {
    document.body.innerHTML = ""
    const { audio, video } = playLightsaberFX()

    expect(audio).toBeNull()
    expect(video).toBeNull()
  })
})
