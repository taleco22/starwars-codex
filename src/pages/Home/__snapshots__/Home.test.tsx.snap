// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.tsx render component correctly 1`] = `
<div>
  <main
    class="sc-AxjAm sc-AxhUy LnXBH"
  >
    <img
      alt="Star Wars"
      class="sc-fzozJi eAHfzr"
      src="[object Object]"
    />
    <div
      class="sc-AxmLO jzjKvw"
    >
      <button
        class="sc-AxhCb sc-Axmtr gFtPZS"
      >
        Start
      </button>
    </div>
  </main>
</div>
`;
